sync-core/src/main/java/com/cloudant/http/CookieFilter.java

```diff
@@ -0,0 +1,118 @@
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by Rhys Short on 15/05/15.
+ *
+ * Adds cookie authentication support to http requests.
+ *
+ * It does this by adding the cookie header for CouchDB
+ * using request filtering pipeline in {@link HttpConnection}.
+ *
+ * If a response has a response code of 401, it will fetch a cookie from
+ * the server using provided credentials and tell {@link HttpConnection} to reply
+ * the request by setting {@link HttpConnectionFilterContext#replayRequest} property to true.
+ *
+ *
+ */
+public  class CookieFilter implements HttpConnectionRequestFilter, HttpConnectionResponseFilter {
+
+
+    final String cookieRequestBody;
+    private String cookie = null;
+
+    /**
+     * Constructs a cookie filter.
+     * @param username The username to use when getting the cookie
+     * @param password The password to use when getting the cookie
+     */
+    public CookieFilter(String username, String password){
+        Map<String,String> cookieRequestMap = new HashMap<String, String>();
+        cookieRequestMap.put("name",username);
+        cookieRequestMap.put("password",password);
+        JSONHelper helper = new JSONHelper();
+        cookieRequestBody = helper.toJson(cookieRequestMap);
+    }
+
+
+    @Override
+    public HttpConnectionFilterContext filterRequest(HttpConnectionFilterContext context) {
+
+        HttpURLConnection connection = context.connection.getConnection();
+
+        if(cookie == null){
+            cookie = getCookie(connection.getURL());
+        }
+        connection.setRequestProperty("Cookie",cookie);
+
+
+        return context;
+
+    }
+
+    @Override
+    public HttpConnectionFilterContext filterResponse(HttpConnectionFilterContext context) {
+        HttpURLConnection connection = context.connection.getConnection();
+        if (context.connection.getResponseCode() == 401) {
+            //we need to get a new cookie
+            cookie = getCookie(connection.getURL());
+            context.replayRequest = true;
+            connection.setRequestProperty("Cookie", cookie);
+
+            context = new HttpConnectionFilterContext(context);
+        }
+        return context;
+
+        }
+
+     private String getCookie(URL url){
+
+        try {
+            URL sessionURL = new URL(String.format("%s://%s:%s/_session",
+                    url.getProtocol(),
+                    url.getHost(),
+                    url.getPort()));
+
+            HttpConnection conn = Http.POST(sessionURL, "application/json");
+            conn.setRequestBody(cookieRequestBody.getBytes("UTF-8"));
+            String cookieHeader = conn.execute().getConnection().getHeaderField("Set-Cookie");
+            return cookieHeader.substring(0,cookieHeader.indexOf(";"));
+
+
+
+        } catch (MalformedURLException e) {
+            e.printStackTrace();
```

Comment:
Need better error logging and handling than this. If we can't get the cookie maybe we should abandon and set replayRequest to false?